### PR TITLE
Add Activity Trace Id Custom Generation Support

### LINKED ISSUE #46704: Support customized Activity trace id generation
## Background and Motivation

The .NET runtime provides APIs to generate Activity, which can be adopted and processed as trace context and eventually emitted to multiple backend vendors by OpenTelemetry .NET SDK for the purpose of tracing. Currently .NET runtime only allows an activity's trace id to be [randomly generated](https://github.com/dotnet/runtime/blob/5f984ae83f0c474e03d7e42a4cec8ab08bba4e8d/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs#L1666) and there is no public API available to override it. While for some backend vendors, trace context with a non-random customized trace id is more preferred. For instance, AWS X-Ray backend will only process trace context with a trace id containing a timestamp. In this case, is it possible for the runtime to support customized activity trace id in Activity class, like adding API similar as Activity.SetTraceId()?

Previous discussion: open-telemetry/opentelemetry-specification#896

*Proposal Edited by @tarekgh*

## Proposed API

``` C#
    Activity.TraceIdGenerationFunction = () => ActivityTraceId.CreateRandom();
    
    ...


    using (Activity activity = source.StartActivity("OperationName"))
    {
        Console.WriteLine(activity?.TraceId);
    }
```

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/runtime/issues/46704

### CODE REVIEW

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 1]
```diff
@@ -832,6 +832,18 @@ public ActivitySpanId ParentSpanId
             }
         }
 
+        /// <summary>
+        /// When starting an Activity which not having any parent context, the Trace Id will automatically generated using random numbers.
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:836]:
```suggestion
        /// When starting an Activity which does not have a parent context, the Trace Id will automatically be generated using random numbers.
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 2]
```diff
@@ -832,6 +832,18 @@ public ActivitySpanId ParentSpanId
...
+        /// TraceIdGenerator can be used to generate the Trace Id instead of using the default Trace Id generation which using the random numbers.
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:837]:
```suggestion
        /// TraceIdGenerator can be used to override the runtime's default trace id generation algorithm.
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 3]
```diff
@@ -832,6 +832,18 @@ public ActivitySpanId ParentSpanId
...
+        /// </summary>
+        /// <remarks>
+        /// - TraceIdGenerator needs to be set only if the default Trace Id generation is not enough for the app scenario.
+        /// - When setting TraceIdGenerator, ensure it is performant enough to avoid any slowness in the Activity starting operation.
+        /// - If TraceIdGenerator is set multiple times, the last set will be the one used for the Trace Id generation.
+        /// - Setting TraceIdGenerator to null will enable back the default Trace Id generation.
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:843]:
```suggestion
        /// - Setting TraceIdGenerator to null will re-enable the default Trace Id generation algorithm.
```

**Review by noahfalk — APPROVED** [review]
LGTM, suggested a few adjustments in the comment text. Thanks @tarekgh!

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 4]
```diff
@@ -1106,7 +1118,8 @@ private void GenerateW3CId()
             {
                 if (!TrySetTraceIdFromParent())
                 {
-                    _traceId = ActivityTraceId.CreateRandom().ToHexString();
+                    Func<ActivityTraceId>? traceIdGenerator = TraceIdGenerator;
+                    _traceId = traceIdGenerator == null ? ActivityTraceId.CreateRandom().ToHexString() : traceIdGenerator().ToHexString();
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:1122]: Nit:
```suggestion
                    Func<ActivityTraceId>? traceIdGenerator = TraceIdGenerator;
                    ActivityTraceId id = traceIdGenerator == null ? ActivityTraceId.CreateRandom() : traceIdGenerator();
                    _traceId = id.ToHexString();
```

**Review by stephentoub — APPROVED** [review]

**Comment by tarekgh** [comment]
The CI leg Windows x64 Release failure is tracked by https://github.com/dotnet/runtime/issues/48169

**Comment by tarekgh** [comment]
Created the issue https://github.com/dotnet/runtime/issues/48198 to track `runtime (Libraries Test Run release coreclr windows x64 Debug)` failure.

**Pull request merged**